### DISCUSSION TITLE
New user having issues rebuilding libmesh

### DISCUSSION THREAD

**Discussion by ParadoxPrism117** [idaholab/moose#28589] [Q&A Getting Started] [answered]
### Check these boxes if you have followed the posting rules.

- [X] I have consulted the [Posting Guidelines](https://github.com/idaholab/moose/discussions/18270).
- [X] I have searched the [Discussions Forum](https://github.com/idaholab/moose/discussions) and [MOOSE Framework Troubleshooting](https://mooseframework.inl.gov/help/troubleshooting.html) and have not found what I was looking for
- [X] Q&A Getting Started is the most appropriate category for my question (trouble installing, beginner user, ...)

### Issue or question about MOOSE

Hi, first time poster and new user to MOOSE. I am trying to install MOOSE and attempted to perform the update/rebuild of libmesh when I encountered an error finding/linking to a "z library". The error provided a suggestion of disabling two applications, but it still didn't build correctly even after trying that. Is there something missing I need to install or disable?

### (Optional) code in question / simulation log / errors

```shell
configure: error: Can't find or link to the z library. Turn off netCDF-4 and      DAP clients with --disable-netcdf-4 --disable-dap, or see config.log for errors.
configure: error: ../../../../contrib/netcdf/netcdf-c-4.6.2/configure failed for contrib/netcdf/netcdf-c-4.6.2
Running make -j 1...
make: *** No targets specified and no makefile found.  Stop.
```

### Encountering Errors? Please include diagnostic output

```pre
####################################################################################################
Influential Environment Variables

CONDA_CHANNEL=https://conda.software.inl.gov/public
CONDA_DEFAULT_ENV=base
CONDA_EXE=/home/kzimins/miniforge/bin/conda
CONDA_PREFIX=/home/kzimins/miniforge
CONDA_PROMPT_MODIFIER=(base) 
CONDA_PYTHON_EXE=/home/kzimins/miniforge/bin/python
CONDA_SHLVL=1
CURL_CA_BUNDLE=
MOOSE_JOBS=6
PATH=/home/kzimins/miniforge/bin:/home/kzimins/miniforge/condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/snap/bin
REQUESTS_CA_BUNDLE=
SSL_CERT_FILE=
####################################################################################################
Compiler(s) (CC CXX FC F77 F90):

CC	not set
CXX	not set
FC	not set
F77	not set
F90	not set

FAIL: One or more compiler environment variables not set, or set incorrectly
####################################################################################################
Python Sanity Checks

/usr/bin/env python3 --version; (reporting as: Python 3.12.5) matches
which python3 python;

/home/kzimins/miniforge/bin/python3 --version; == Python 3.12.5
/home/kzimins/miniforge/bin/python --version; == Python 3.12.5

OK
####################################################################################################
Python Modules (TestHarness, run-ability)


WARNING: missing module(s): pyaml jinja2 

Either install the above packages, or perhaps you have yet
to activate the moose environment: conda activate moose

Missing 'warning' Python modules may cause miscellaneous runtime issues.

WARNING: One or more Python issues present.
####################################################################################################
MOOSE Repository/Conda Version Checks

WARNING: Unable to run Conda tests due to missing Python modules


checks FAILED
```

**Comment by GiudGiud** [comment]
Hello

You need to install zlib.
You can get it from conda fairly easily `conda install zlib`

**Reply by ParadoxPrism117** [reply]
Does zlib need to be installed in the same folder?

**Reply by ParadoxPrism117** [reply]
I installed zlib but I am still having the same issue. Do I need to attempt a clean install of MOOSE?

**Reply by GiudGiud** [reply]
> Does zlib need to be installed in the same folder?

no should be fine anywhere  as long as it s found in the LIBRARY_PATH

> Do I need to attempt a clean install of MOOSE?

probably not. let s see the config.log for libmesh

**Reply by ParadoxPrism117** [reply]
[config.log](https://github.com/user-attachments/files/16937821/config.log)

**Reply by ParadoxPrism117** [reply]
I attached the config.log as a file because it was too much text to paste into a reply

**Reply by milljm** [reply]
Looks like netcdf failed to configure/build. Can you navigate where it says its building it in the log:
```pre
/home/kzimins/projects/moose/libmesh/build/contrib/netcdf/netcdf-c-4.6.2
```
and report what the `configure.log` says there? It may be named `config.log` as well. Or some other ".log" file.

**Reply by ParadoxPrism117** [reply]
sure, here is the netcdf config.log: 
[config.log](https://github.com/user-attachments/files/16945160/config.log)

**Reply by ParadoxPrism117** [reply]
It mentions on line 530 that it encounters a fatal error with a "curl.h" file/directory?

**Reply by GiudGiud** [reply]
I think the error is still the same, towards the end
This curl.h error you can solve by installing libcurl-devel

**Reply by milljm** [reply]
This still seems to be a zlib error:
```pre
configure:18515: error: Can't find or link to the z library. Turn off netCDF-4 and      DAP clients with --disable-netcdf-4 --disable-dap, or see config.log for errors.
```

I think installing zlib using Conda only works if you also use Conda's compilers. What I mean by that, is that the following link line:

```pre
configure:18486: mpicc -o conftest  -I/home/kzimins/projects/moose/scripts/../petsc/include \
-I/home/kzimins/projects/moose/scripts/../petsc/arch-moose/include \
-I/home/kzimins/projects/moose/petsc/arch-moose/include \
-I/home/kzimins/projects/moose/petsc/include \
-I/home/kzimins/projects/moose/petsc/arch-moose/include    conftest.c -lz  \
-Wl,-rpath,/home/kzimins/projects/moose/petsc/arch-moose/lib \
-L/home/kzimins/projects/moose/petsc/arch-moose/lib \
-Wl,-rpath,/usr/lib/gcc/x86_64-linux-gnu/11 \
-L/usr/lib/gcc/x86_64-linux-gnu/11 -lpetsc -lHYPRE -ldmumps -lmumps_common \
-lpord -lpthread -lstrumpack -lscalapack -lsuperlu_dist -lflapack -lfblas \
-lptesmumps -lptscotchparmetisv3 -lptscotch -lptscotcherr -lesmumps -lscotch \
-lscotcherr -lparmetis -lmetis -lhdf5_hl -lhdf5 -lm -lX11 -lmpichfort -lmpich \
-lgfortran -lm -lgfortran -lm -lgcc_s -lquadmath -lstdc++ -lrt -L/lib -lhdf5 \
-Wl,-rpath,/lib   >&5
```
Contains no paths to your Conda installation. The following:
```pre
-L/usr/lib/gcc/x86_64-linux-gnu/11
```
This is the bit I see in the above link line that includes a path to your compiler. If this were Conda supplied compilers, it would report: `-L/home/kzimins/miniforge/envs/moose/lib`, and by extension, would include the path necessary to Conda installed z library.

I am sure there is a reason you are not wanting to install the MOOSE Conda packages, so I won't press. So instead, try this:
```bash
export LDFLAGS="-L/home/kzimins/miniforge/lib $LDFLAGS"
export METHODS=opt MOOSE_JOBS=6  # this will help speed up the process
./update_and_rebuild_libmesh.sh
```
However, because there will be plenty of other libraries in this location, I can't guarantee success. There may be additional issues. If so, it may be better if you `apt install zlib-dev` or build the z library manually, and add the above `-L` flag pointing to the sole one item you manually create.

**Reply by ParadoxPrism117** [reply]
I have no problem with using the MOOSE Conda packages as far as I am aware, how do I go about installing them?

**Reply by GiudGiud** [reply]
are you using a personal machine or a computing cluster? We recommend conda packages for the former

**Reply by milljm** [reply, marked as answer]
Out of curiosity, what instructions were you following? I'd like to understand if we led you astray somehow!

If you are not against installing the MOOSE Conda packages, this all becomes much simpler:
```bash
conda create -n moose moose-dev mpich
conda activate moose
cd ~/projects/moose

# Lets clean out all previous failed builds for good measure
git clean -Xfd; git submodule foreach --recursive git clean -Xfd

cd test
make -j 6
./run_tests -j 6
```
(6, or however many CPU cores you have). No need to build PETSc, libMesh, or WASP (our packages supply these pre-built).

**Reply by ParadoxPrism117** [reply]
> are you using a personal machine or a computing cluster? We recommend conda packages for the former

local machine

**Reply by ParadoxPrism117** [reply]
> Out of curiosity, what instructions were you following? I'd like to understand if we led you astray somehow!
> 
> If you are not against installing the MOOSE Conda packages, this all becomes much simpler:
> 
> ```shell
> conda create -n moose moose-dev mpich
> conda activate moose
> cd ~/projects/moose
> 
> # Lets clean out all previous failed builds for good measure
> git clean -Xfd; git submodule foreach --recursive git clean -Xfd
> 
> cd test
> make -j 6
> ./run_tests -j 6
> ```
> 
> (6, or however many CPU cores you have). No need to build PETSc, libMesh, or WASP (our packages supply these pre-built).

The gcc compiler I have on there right now is from some previous linux-based software I was using. This is what I was trying to follow to get MOOSE installed: https://mooseframework.inl.gov/getting_started/installation/install_moose.html

**Reply by GiudGiud** [reply]
These are the better instructins for your case
https://mooseframework.inl.gov/getting_started/installation/conda.html

**Reply by ParadoxPrism117** [reply]
I made it to the part when I was supposed to build MOOSE and then this happened:

Compiling C++ (in opt mode) /home/kzimins/projects/moose/framework/build/unity_src/meshmodifiers_Unity.C...
{standard input}: Assembler messages:
{standard input}:738153: Warning: end of file not at end of a line; newline inserted
{standard input}:738468: Error: expecting operand after ','; got nothing
{standard input}: Error: open CFI at the end of file; missing .cfi_endproc directive
x86_64-conda-linux-gnu-c++: fatal error: Killed signal terminated program cc1plus
compilation terminated.
make: *** [/home/kzimins/projects/moose/framework/build.mk:150: /home/kzimins/projects/moose/framework/build/unity_src/utils_Unity.x86_64-conda-linux-gnu.opt.lo] Error 1
make: *** Waiting for unfinished jobs....

**Reply by jmeier** [reply]
@ParadoxPrism117, how many jobs do you use (the j-parameter after make, e.g. make -j4"). Try to reduce the number.

**Reply by ParadoxPrism117** [reply]
I reduced it to 2 and it worked, thanks for the help!
